src/librustc_mir/interpret/snapshot.rs

```diff
@@ -1,31 +1,66 @@
-//! This module contains the machinery necessary to detect infinite loops
-//! during const-evaluation by taking snapshots of the state of the interpreter
-//! at regular intervals.
-
-// This lives in `interpret` because it needs access to all sots of private state.  However,
-// it is not used by the general miri engine, just by CTFE.
-
+//! Capture and compare snapshots of the compile-time interpreter state to detect when a program
+//! will loop infinitely.
+//!
+//! This lives in `interpret` because it needs access to all sorts of private state.  However,
+//! it is not used by the general miri engine, just by CTFE.
+//!
+//! # Comparing interpreter state
+//!
+//! If we observe the same interpreter state at two different points in time during const-eval, we
+//! can be certain that the program is in an infinite loop. For `CompileTimeInterpreter`, the
+//! relevant bits of interpreter state are the stack and the heap. We refer to a copy of this state
+//! as a "snapshot".
+//!
+//! While comparing stacks between snapshots is straightforward, comparing heaps between snapshots
+//! is not. This is because `AllocId`s, which determine the target of a pointer, are frequently
+//! thrown away and recreated, even if the actual value in memory did not change. Consider the
+//! following code:
+//!
+//! ```rust,ignore(const-loop)
+//! const fn inf_loop() {
+//!     loop {
+//!         let arr = [0isize; 4];
+//!         let parr = &arr;
+//!         if parr[0] != 0 {
+//!             break;
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! Although this program will loop indefinitely, a new `AllocId` will be created for the value
+//! pointed to by `arr` at each iteration of the loop. A naive method for comparing snapshots, one
+//! that considered the numeric value of `AllocId`s, would cause this class of infinite loops to be
+//! missed. See [#52475](https://github.com/rust-lang/rust/issues/52475).
+//!
+//! Instead, we assign a new, linear index, one that is used only while comparing snapshots, to
+//! each allocation. This index, hereby referred to as the "DFS index" is the order that each
+//! allocation would be encountered during a depth-first search of all allocations reachable from
+//! the stack via a pointer. See below for an example of this numbering scheme.
+//!
+//! ```text
+//! Stack pointers:    x  x    (DFS starts with leftmost pointer on top of stack)
+//!                    |  |
+//! Heap allocs:     1 o  o 3
+//!                    | /|
+//!                    |/ |
+//!                  2 o  o 4
+//! ```
+//!
+//! Instead of comparing `AllocId`s between snapshots, we first map `AllocId`s in each snapshot to
+//! their DFS index, then compare those instead.
```

Comment:
That makes sense. However, what I would have expected then is to see `Pointer<Tag, DfsIndex>` instead of `Pointer<Tag, AllocId>`, and then one would compare those. What I still don't get is why we have a way to compare pointers *ignoring* the `AllocId`. That means we must have a magic scheme somewhere to figure out all the `AllocId` we ignored and compare their `DfsIndex` instead? That sounds quite round-about and fragile, on first sight. (I didn't read the code, though.)